File: LeetCode/LeetCode/Problems/10/Problem1095.cs
public class Problem1095
{
    public int FindInMountainArray(int target, MountainArray mountainArr)
    {
        int length = mountainArr.Length();

        int left = 1;
        int right = length - 2;
        while (left != right)
        {
            int mid = (left + right) / 2;
            if (mountainArr.Get(mid) < mountainArr.Get(mid + 1))
            {
                left = mid + 1;
            }
            else
            {
                right = mid;
            }
        }
        int peakIndex = left;

        left = 0;
        right = peakIndex;
        while (left != right)
        {
            int testIndex = (left + right) / 2;
            if (mountainArr.Get(testIndex) < target)
            {
                left = testIndex + 1;
            }
            else
            {
                right = testIndex;
            }
        }

        if (mountainArr.Get(left) == target)
        {
            return left;
        }

        left = peakIndex + 1;
        right = length - 1;
        while (left != right)
        {
            int testIndex = (left + right) / 2;
            if (mountainArr.Get(testIndex) > target)
            {
                left = testIndex + 1;
            }
            else
            {
                right = testIndex;
            }
        }

        if (mountainArr.Get(left) == target)
        {
            return left;
        }

        return -1;
    }

    public interface MountainArray
    {
        public int Get(int index);
        public int Length();
    }
}

File: leetcode_rust/src/problems/problem_1095.rs
pub fn find_in_mountain_array(target: i32, mountain_arr: &MountainArray) -> i32 {
    let length = mountain_arr.length();

    let mut left = 1;
    let mut right = length - 2;
    while left != right {
        let mid = (left + right) / 2;
        if mountain_arr.get(mid) < mountain_arr.get(mid + 1) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    let peak_index = left;

    left = 0;
    right = peak_index;
    while left != right {
        let test_index = (left + right) / 2;
        if mountain_arr.get(test_index) < target {
            left = test_index + 1;
        } else {
            right = test_index;
        }
    }

    if mountain_arr.get(left) == target {
        return left;
    }

    left = peak_index + 1;
    right = length - 1;
    while left != right {
        let test_index = (left + right) / 2;
        if mountain_arr.get(test_index) > target {
            left = test_index + 1;
        } else {
            right = test_index;
        }
    }

    if mountain_arr.get(left) == target {
        return left;
    }

    return -1;
}

struct MountainArray;

impl MountainArray {
    fn get(_: i32) -> i32 {
        0
    }
    fn length() -> i32 {
        0
    }
}
